invokeai/backend/stable_diffusion/diffusion/custom_attention.py

```diff
@@ -291,6 +312,35 @@ def run_attention_normal(
 
         return hidden_states
 
+    def run_attention_xformers(
```

Comment:
Because it's will be hard to address multihead attention in sliced attention.
Also because I left it all as it in diffusers almost completely.
Also sliced attention in case of single loop(no slice) allocates 2x memory because it preallocates tensor for result.

Reply:
> Because it's will be hard to address multihead attention in sliced attention. 

I don't follow, what is hard about it? We handle the batch/head dimensions the same way for sliced as for normal and xformers. Maybe there is a slight performance benefit to not calling `head_to_batch_dim(...)` for SDP? That would be a reason to keep both.

> Also sliced attention in case of single loop(no slice) allocates 2x memory because it preallocates tensor for result.

True. In context it's pretty small - on the order of 10MB. But... maybe that costs a non-negligible amount of time.

I'm fine with leaving this for now. It would take more performance testing to confirm that it's actually a good idea.

Reply:
Jeez, you really want to reimplement this all.
`torch-sdp` enabled more optimizations for `(batch, num_heads, seq_len, head_dim)` inputs.
And in other attentions we throw away this info and combine heads and batches.

Reply:
Also about size - hidden_states sizes(fp16):
1024x1024

| resolution  | size |
| ------------- | ------------- |
| 512x512  | 320 mb  |
| 1024x1024 | 1.25 gb  |
| 2048x2048 | 5 gb |
| 4096x4096 | 20gb |

Reply:
After all, I reimplemented it, so, now you can look at it again.